homeassistant/components/onkyo/media_player.py

```diff
@@ -108,248 +186,224 @@
-            self._attributes.pop(ATTR_AUDIO_INFORMATION, None)
-            self._attributes.pop(ATTR_VIDEO_INFORMATION, None)
-            self._attributes.pop(ATTR_PRESET, None)
-            self._attributes.pop(ATTR_VIDEO_OUT, None)
-            return
+    # Discover what zones are available for the avr by querying the power.
+    # If we get a response for the specific zone, it means it is available.
+    for zone in ZONES:
+        avr.query_property(zone, "power")
 
-        volume_raw = self.command("volume query")
-        mute_raw = self.command("audio-muting query")
-        current_source_raw = self.command("input-selector query")
-        # If the following command is sent to a device with only one HDMI out,
-        # the display shows 'Not Available'.
-        # We avoid this by checking if HDMI out is supported
-        if self._hdmi_out_supported:
-            hdmi_out_raw = self.command("hdmi-output-selector query")
-        else:
-            hdmi_out_raw = []
-        preset_raw = self.command("preset query")
-        if self._audio_info_supported:
-            audio_information_raw = self.command("audio-information query")
-        if self._video_info_supported:
-            video_information_raw = self.command("video-information query")
-        if not (volume_raw and mute_raw and current_source_raw):
-            return
+    # Add the main zone to active_zones, since it is always active
+    main_entity = OnkyoAVR(avr, name, sources, "main", max_volume)
+    active_zones["main"] = main_entity
 
-        sources = _parse_onkyo_payload(current_source_raw)
+    @callback
+    def close_avr(_event):
+        avr.close()
 
-        for source in sources:
-            if source in self._source_mapping:
-                self._current_source = self._source_mapping[source]
-                break
-            self._current_source = "_".join(sources)
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, close_avr)
 
-        if preset_raw and self._current_source.lower() == "radio":
-            self._attributes[ATTR_PRESET] = preset_raw[1]
-        elif ATTR_PRESET in self._attributes:
-            del self._attributes[ATTR_PRESET]
+    async_add_entities([main_entity])
 
-        self._muted = bool(mute_raw[1] == "on")
-        #       AMP_VOL/MAX_RECEIVER_VOL*(MAX_VOL/100)
-        self._volume = volume_raw[1] / (
-            self._receiver_max_volume * self._max_volume / 100
-        )
 
-        self._parse_audio_information(audio_information_raw)
-        self._parse_video_information(video_information_raw)
+class OnkyoAVR(MediaPlayerEntity):
+    """Entity reading values from Onkyo AVR protocol."""
 
-        if not hdmi_out_raw:
+    def __init__(self, avr, name, sources, zone, max_volume):
+        """Initialize entity with transport."""
+        super().__init__()
+        self._avr = avr
+        self._name = f"{name} {ZONES[zone] if zone != 'main' else ''}"
+        self._unique_id = None
```

Comment:
Can we not use the receivers mac address here to give a good unique id?

Reply:
I have a PR on the pyeiscp library that adds a unique identifier to the AVR object. So as soon as that gets merged, I can update the code here to use that identifier!